### PR TITLE
Adding Classic Richardson WGS Solver

### LINKED ISSUE #22: classic source iteration
The traditional SI (classic Richardson) should be re-implemented in the code. It was removed a long time ago but it is a useful technique

Special attention should be paid to lines such as in [modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h](https://github.com/Open-Sn/opensn/blob/53422091a1d50431ad044c65224bf1691aad9032/modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h#L15)
```
case IterativeMethod::CLASSICRICHARDSON:
case IterativeMethod::CLASSICRICHARDSON_CYCLES:
case IterativeMethod::KRYLOV_RICHARDSON:
case IterativeMethod::KRYLOV_RICHARDSON_CYCLES:
    return "richardson";
```

When krylov_richardson is used, an extra sweep is needed because it comes out of a KSP solve. For classic richardson, this will not be the case.

### PR DESCRIPTION
**Pull request by wdhawkins**
This PR adds a WGS solver based on Classic Richardson. This class is not derived from WGSLinearSolver since it does not depend on PETSc with the exception of whatever the preconditioners do. Resolves #22.

Moving forward, it might make sense to rename this to WGSClassicRicardsonSolver and rename the current WGSLinearSolver to WGSPETScSolver or something similar.  Thoughts?

### CODE REVIEW

**Comment by wdhawkins** [comment]
@zhardy-lanl it seems to me that all I should need to do to get preconditioners working is to add the following to `classic_richardson.cc` after `gs_context_ptr->ApplyInverseTransportOperator(scope);`
```
gs_context_ptr->ApplyInverseTransportOperator(scope);
 
// Apply WGDSA
if (groupset.apply_wgdsa_)
{
  auto& phi_new_local = lbs_solver.PhiNewLocal();
  std::vector<double> delta_phi_local;
  lbs_solver.AssembleWGDSADeltaPhiVector(groupset, phi_new_local, delta_phi_local);
  groupset.wgdsa_solver_->Assemble_b(delta_phi_local);
  groupset.wgdsa_solver_->Solve(delta_phi_local);
  lbs_solver.DisAssembleWGDSADeltaPhiVector(groupset, delta_phi_local, phi_new_local);
}
  
// Apply TGDSA
if (groupset.apply_tgdsa_)
{
  auto& phi_new_local = lbs_solver.PhiNewLocal();
  std::vector<double> delta_phi_local;
  lbs_solver.AssembleTGDSADeltaPhiVector(groupset, phi_new_local, delta_phi_local);
  groupset.tgdsa_solver_->Assemble_b(delta_phi_local);
  groupset.tgdsa_solver_->Solve(delta_phi_local);
  lbs_solver.DisAssembleTGDSADeltaPhiVector(groupset, delta_phi_local, phi_new_local);
}
```

But this doesn't seem to work... it goes unstable. Any thoughts?

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (andrsd) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:18]: 🗒️  If #328 goes in before this PR, then this will need to be removed before merging

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (andrsd) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:20]: This looks like a stray comment, this is a c-tor.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (andrsd) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:37]: The `;` is not needed and is most likely causing this weird formatting. See line 32 for comparison. 

So remove the `;` and make the same change for the functions below as well.

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:37]: Looks like I got a bit `;` happy. Fixed!

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (ragusa) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:77]: at first, I was confused about these 2 routines. Would you consider changing the name of the first one to reflect what PW is computed by adding `Phi`, so that it reads `ComputePointwisePhiChange` ?

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: This is the within-groupset source function signature (see `modules/linear_boltzmann_solvers/lbs_solver/source_functions.`). It also mirrors what is currently used in `WGSContext` (see `modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/wgs_context.h`).

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua:115]: I'll remove it. I need to go through a remove a lot of non-functional comments in the input decks.

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua:151]: The classic Richardson inputs are copies of existing inputs.  This baggage just came with them. I'll remove this put up a PR that cleans up all of the inputs.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:34]: Prototypes without a definition are statements and always require a semicolon. C++ does not require that you put a semicolon after method declarations (but you can).

File: modules/linear_boltzmann_solvers/discrete_ordinates_solver/iterative_methods/sweep_wgs_context.cc
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_solver/iterative_methods/sweep_wgs_context.cc:158]: Fixed.

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_5_cycles_2_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_5_cycles_2_classic_richardson.lua:102]: Fixed.

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua:115]: Fixed.

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua:151]: Fixed.

File: test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_steady/transport_3d_2_unstructured_classic_richardson.lua:26]: Yes. I've also converted one of the existing 3D ParMETIS inputs to Classic Richardson and added it to the regression tests.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:77]: Fixed.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (ragusa) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: @wdhawkins can we talk about this soon this coming week?

Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: Yes, let's discuss this on Wednesday. It's probably a good idea to have @zhardy-lanl there. He probably knows more about the history of this than I do.

**Review by andrsd — APPROVED** [review]

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: This was perhaps a sloppy code design on my part from before. I think there is a pretty straightforward fix to remove this. I'll put up a PR for this soon. This is part of the reason I originally brought up removing densities from `LBSSolver` about a month ago.

Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: That sounds good to me. In that case, can I get approvals for this PR since the design matches that of the other WGS solvers?

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (zhardy-lanl) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:59]: If I am reading this correctly, now there is `psi_new_local` in `LBSSolver` and then `psi_new` and `psi_old` per groupset? Do the PETSC solvers converge on angular fluxes? Should they? If not, why are we doing this here?

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (zhardy-lanl) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:61]: This feels like it should belong in `DiscreteOrdinatesSolver` to be honest. However, since this class does not necessarily have an instance of that (even through contexts), I am not sure how to make that work.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:49]: Working on it now

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h
Comment (zhardy-lanl) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h:35]: I understand this is to return a PETSC iterative method name, but this still looks funny to me. Since the default return value seems to be `""`, is this necessary here?

File: modules/linear_boltzmann_solvers/lbs_solver/lbs_solver.h
Comment (zhardy-lanl) [inline modules/linear_boltzmann_solvers/lbs_solver/lbs_solver.h:66]: This should be moved to more like routines instead of being embedded within the getters/setters. Maybe near the other post-iteration routines.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (zhardy-lanl) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:1]: Would it have been a total headache to implement this within `WGSSolver` with a few if statements to avoid the PETSc stuff? I ask particularly because `LinearSolver` has PETSc stuff build into it already.

File: test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua
Comment (zhardy-lanl) [inline test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua:1]: Just a note that there is mixed use of `cr` and `classic_richardson` in the new test file names. I'm not sure which is better, but consider using the same notation throughout.

**Review by zhardy-lanl — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:59]: I can probably get rid of `psi_new_` here and use `psi_new_local` in `LBSSolver` instead. I'll see about making that change. The PETSc solvers do converge on angular fluxes... which is the appropriate thing to do.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:61]: I disagree. This is specific to this WGS solver and uses private member variables. I don't see that it will ever be used anywhere else.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h:35]: This is solely to prevent a compiler warning that not all values are parsed in the switch.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:1]: Yes, I had really hoped to do that, but it turns out to be considerably messier. I'm hoping that we can simplify the iterative algorithm infrastructure a bit to make that a reality.

File: test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua:1]: Yeah, I went with `cr` there just to prevent the file name from being so long. I'll work on making it consistent.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/lbs_iterative_methods.h:35]: Couldn't you use `default:` here then?

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:61]: That is fair. The same argument be made for `ComputePointwisePhiChange` which was implemented in `LBSSolver`? Where should that live instead?

Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:61]: Yes, the same argument could be made for `ComputePointwisePhiChange`. I'll make it a private method. The reason I originally had it in `LBSSovler` was that I expected the functionality to overlap with the pointwise phi check for the refactored AGS solver. But, it turned out that trying to reuse that code would result in a slew of unnecessary collective MPI operations.

Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:61]: I won't hold things up on this point, but it could be worth considering defining these methods as standalone (i.e. taking an instance of `LBSSolver` and the new and old vectors) somewhere in the LBS module to avoid these issues, if that is feasible.

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc
Comment (andrsd) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc:18]: Just a comment that #328 is in, so do not forget to remove this before merging...

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:59]: I've opted against using `psi_new_local` in `LBSSolver` for this purpose for a couple of reasons. First, it would prevent the use of classic Richardson with time-dependent or adjoint solvers. Maybe not prevent, but it would be very inefficient. Second, I'm not sure we want to have psi at the level of `LBSSolver` or even at the discrete ordinates solver level. It seems to me this should be a private member for each individual solver (adjoint, time-dependent, etc.). I think we need to look more closely at how we want to handle storing psi as part of the refactoring of `LBSSolver`.

**Review by ragusa — APPROVED** [review]

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (andrsd) [inline modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:20]: This comment should be now:
```
/**
 *
 */
```

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h:18]: Fixed.

File: test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua
Comment (wdhawkins) [reply test/modules/linear_boltzmann_solvers/transport_keigen/keigenvalue_transport_3d_openmcxs_UO2_cr.lua:1]: Fixed.

**Review by zhardy-lanl — APPROVED** [review]
Thanks for the changes! Some of the more fine grained details should be hashed out in future PRs when we change LBS solver.

**Comment by wdhawkins** [comment]
> Thanks for the changes! Some of the more fine grained details should be hashed out in future PRs when we change LBS solver.

Yes, I really would like to see the convergence checks pulled out into stand-alone routines. I have an upcoming PR that does something similar for LBS io, so it might be a pattern that we can use for convergence checking.

**Review by andrsd — APPROVED** [review]

**Pull request merged**
